http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java

```diff
@@ -0,0 +1,243 @@
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.TrustManagerFactory;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.http.Protocol;
+import software.amazon.awssdk.http.nio.netty.SdkEventLoopGroup;
+import software.amazon.awssdk.http.nio.netty.internal.http2.HttpOrHttp2ChannelPool;
+import software.amazon.awssdk.utils.Logger;
+
+/**
+ * Implementation of {@link SdkChannelPoolMap} that awaits channel pools to be closed upon closing.
+ */
+@SdkInternalApi
+public final class AwaitCloseChannelPoolMap extends SdkChannelPoolMap<URI, AwaitCloseChannelPoolMap.SimpleChannelPoolAwareChannelPool> {
+
+    private static final Logger log = Logger.loggerFor(AwaitCloseChannelPoolMap.class);
+
+    private final SdkChannelOptions sdkChannelOptions;
+    private final SdkEventLoopGroup sdkEventLoopGroup;
+    private final NettyConfiguration configuration;
+    private final Protocol protocol;
+    private final long maxStreams;
+    private final SslProvider sslProvider;
+
+    private AwaitCloseChannelPoolMap(Builder builder) {
+        this.sdkChannelOptions = builder.sdkChannelOptions;
+        this.sdkEventLoopGroup = builder.sdkEventLoopGroup;
+        this.configuration = builder.configuration;
+        this.protocol = builder.protocol;
+        this.maxStreams = builder.maxStreams;
+        this.sslProvider = builder.sslProvider;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    protected SimpleChannelPoolAwareChannelPool newPool(URI key) {
+        SslContext sslContext = sslContext(key.getScheme());
+        Bootstrap bootstrap =
+            new Bootstrap()
+                .group(sdkEventLoopGroup.eventLoopGroup())
+                .channelFactory(sdkEventLoopGroup.channelFactory())
+                .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, configuration.connectTimeoutMillis())
+                // TODO run some performance tests with and without this.
+                .remoteAddress(key.getHost(), key.getPort());
+        sdkChannelOptions.channelOptions().forEach(bootstrap::option);
+
+        AtomicReference<ChannelPool> channelPoolRef = new AtomicReference<>();
+        ChannelPipelineInitializer handler =
+            new ChannelPipelineInitializer(protocol, sslContext, maxStreams, channelPoolRef, configuration, key);
+
+        BetterSimpleChannelPool simpleChannelPool = new BetterSimpleChannelPool(bootstrap, handler);
+
+        channelPoolRef.set(wrapSimpleChannelPool(bootstrap, simpleChannelPool));
+        return new SimpleChannelPoolAwareChannelPool(simpleChannelPool, channelPoolRef.get());
+    }
+
+    @Override
+    public void close() {
+        log.trace(() -> "Closing channel pools");
+        Collection<SimpleChannelPoolAwareChannelPool> channelPools = pools().values();
+        super.close();
```

Comment:
Could something get added between snapshotting the channels in the pools and cleaning up?

Reply:
It's definitely possible but should be very unlikely to happen in my mind. I'm not very concerned about this edge case.

Reply:
What's the impact in the rare case that it does?

Reply:
The impact is the same race condition where the event loop threads get terminated before the newly added channels being closed and it could cause the eventLoop.shutdown to hang.

Reply:
Hmm. Is there any way to avoid that, or no?